gctscript/wrappers/gct/exchange/exchange_test.go

```diff
@@ -0,0 +1,195 @@
+	}
+}
+
+func TestExchange_GetExchange(t *testing.T) {
+	t.Parallel()
+	_, err := exchangeTest.GetExchange(exchName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = exchangeTest.GetExchange("hello world")
+	if err == nil {
+		t.Fatal("unexpected error message received nil")
+	}
+}
+
+func TestExchange_IsEnabled(t *testing.T) {
+	t.Parallel()
+	x := exchangeTest.IsEnabled(exchName)
+	if !x {
+		t.Fatal("expected return to be true")
+	}
+	x = exchangeTest.IsEnabled("fake_exchange")
+	if x {
+		t.Fatal("expected return to be false")
+	}
+}
+
+func TestExchange_Ticker(t *testing.T) {
+	t.Parallel()
+	c := currency.NewPairDelimiter(pairs, delimiter)
+	_, err := exchangeTest.Ticker(exchName, c, assetType)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestExchange_Orderbook(t *testing.T) {
+	t.Parallel()
+	c := currency.NewPairDelimiter(pairs, delimiter)
+	_, err := exchangeTest.Orderbook(exchName, c, assetType)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestExchange_Pairs(t *testing.T) {
+	t.Parallel()
+	_, err := exchangeTest.Pairs(exchName, false, assetType)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = exchangeTest.Pairs(exchName, true, assetType)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestExchange_AccountInformation(t *testing.T) {
+	if !configureExchangeKeys() {
+		t.Skip("no exchange configured test skipped")
+	}
+	tx, err := exchangeTest.AccountInformation(exchName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Log(tx)
+}
+
+func TestExchange_QueryOrder(t *testing.T) {
+	if !configureExchangeKeys() {
+		t.Skip("no exchange configured test skipped")
+	}
+	_, err := exchangeTest.QueryOrder(exchName, orderID)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestExchange_SubmitOrder(t *testing.T) {
+	if !configureExchangeKeys() {
+		t.Skip("no exchange configured test skipped")
+	}
+	tempOrder := &order.Submit{
+		Pair:         currency.NewPairDelimiter(pairs, delimiter),
+		OrderType:    orderType,
+		OrderSide:    orderSide,
+		TriggerPrice: 0,
+		TargetAmount: 0,
+		Price:        orderPrice,
+		Amount:       orderAmount,
+		ClientID:     orderClientID,
+	}
+	_, err := exchangeTest.SubmitOrder(exchName, tempOrder)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestExchange_CancelOrder(t *testing.T) {
+	if !configureExchangeKeys() {
+		t.Skip("no exchange configured test skipped")
+	}
+	_, err := exchangeTest.CancelOrder(exchName, orderID)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func setupEngine() (err error) {
+	engine.Bot, err = engine.NewFromSettings(&settings)
+	if engine.Bot == nil || err != nil {
+		return err
+	}
+	return engine.Bot.Start()
+}
+
+func cleanup() (err error) {
+	err = os.RemoveAll(settings.DataDir)
+	if err != nil {
+		return
+	}
+	return nil
+}
+
+func configureExchangeKeys() bool {
+	ex := engine.GetExchangeByName(exchName).GetBase()
+	ex.SetAPIKeys(exchAPIKEY, exchAPISECRET, exchClientID)
+	ex.SkipAuthCheck = true
+	return ex.ValidateAPICredentials()
```

Comment:
This will always return true if SkipAuthCheck is set to true

Reply:
Based on the logic behind ValidateAPICredentials() this isn't the case 

ValidateAPICredentials() only checks to see if you have keys set/different from default/base64 decode if required 

The SkipAuthCheck was just used as an alternative to 
```
	ex.API.AuthenticatedSupport = true
	ex.API.AuthenticatedWebsocketSupport = true
```
https://github.com/thrasher-corp/gocryptotrader/blob/master/exchanges/exchange.go#L482

As i set the keys after a engine init/config load etc has already happened for these tests

Reply:
Yep, you're correct. I was thinking of AllowAuthenticatedRequest() and I made both those funcs as part of engine 😆